cinderlib/cinderlib.py

```diff
@@ -101,6 +101,11 @@ def __init__(self, volume_backend_name, **driver_cfg):
 
         self._pool_names = tuple(pool['pool_name'] for pool in stats['pools'])
 
+    @staticmethod
+    def driver_options(volume_driver):
+        """Return the driver specific configuration options."""
+        return importutils.import_class(volume_driver).get_driver_options()
+
```

Comment:
I love the idea of being able to report the configuration options for the drivers!!

I think in most cases the caller will want the information for all drivers, and doing it this way would end up with all the drivers loaded in memory, but I still see value in being able to get the options for a specific driver.

That's why I would like to do 2 things:

- Add this information to the data returned by `list_supported_drivers`, which already spawns a different process so we free the memory after we get the info.
- Make this method also use a different process to retrieve the information

If the Cinder patch adds this information to `DriverInfo` in `cinder/interface/util.py` we could get this information directly here!
For the time being I'm going to modify the mapping in `list_supported_drivers` so it returns a dictionary for each entry instead of a single string.  That way it'll be more extensible and allow us to add this information.

Reply:
I'm working on the next patchset that will add the info to the cinder/interface/util.py

Reply:
ok I pushed up a new patchset that includes dumping the options in generate_driver_list.py

http://paste.openstack.org/show/744690/

Reply:
The output looks amazing!!
With that mechanism in place cinderlib will be able to return not only name, description, and type, but also default value and choices for each configuration option!
Which is going to be a game changer for cinderlib, as it will help get seamless integrations in other systems and also assist developers.

So we'll just need to convert driver_options to something we can serialize, so we could replace [the code here](https://github.com/Akrog/cinderlib/blob/master/cinderlib/cinderlib.py#L405):

```
for driver in mapping.values():
    driver.pop('cls', None)
```

With:

```
for driver in mapping.values():
    driver.pop('cls', None)
    options = driver.get('driver_options')
    if options:
        driver['driver_options'] = {
            opt.name: {'help': opt.help, 'type': '%s' % opt.type,
                              'default': opt.default, 'choices': opt.type.choices}
            for opt in options
        }
```

Or something like that, since I don't know if the default and choices are right.